agent/task_context.go

```diff
@@ -191,14 +192,262 @@ func (tc *taskContext) getExecTimeout() time.Duration {
+	if tg == nil {
+		return nil, errors.Errorf("couldn't find task group '%s' in project '%s'", taskGroup, tc.taskConfig.Project.Identifier)
+	}
+
+	return &commandBlock{
+		block:       command.SetupTaskBlock,
+		commands:    tg.SetupTask,
+		timeoutKind: setupTaskTimeout,
+		getTimeout:  tc.getSetupTaskTimeout(tg),
+		canFailTask: tg.SetupTaskCanFailTask,
+	}, nil
+}
+
+// getPost returns a command block containing the post task commands.
+func (tc *taskContext) getPost(taskGroup string) (*commandBlock, error) {
+	if err := tc.taskConfig.Validate(); err != nil {
+		return nil, err
+	}
+
+	if taskGroup == "" {
+		return &commandBlock{
+			block:       command.PostBlock,
+			commands:    tc.taskConfig.Project.Post,
+			timeoutKind: postTimeout,
+			getTimeout:  tc.getPostTimeout,
+			canFailTask: tc.taskConfig.Project.PostErrorFailsTask,
+		}, nil
+	}
+
+	tg := tc.taskConfig.Project.FindTaskGroup(taskGroup)
+	if tg == nil {
+		return nil, errors.Errorf("couldn't find task group '%s' in project '%s'", taskGroup, tc.taskConfig.Project.Identifier)
+	}
+	return &commandBlock{
+		block:       command.TeardownTaskBlock,
+		commands:    tg.TeardownTask,
+		timeoutKind: teardownTaskTimeout,
+		getTimeout:  tc.getTeardownTaskTimeout(tg),
+		canFailTask: tg.TeardownTaskCanFailTask,
+	}, nil
+}
+
+// getSetupGroup returns the setup group for a task group task.
+func (tc *taskContext) getSetupGroup(taskGroup string) (*commandBlock, error) {
+	if err := tc.taskConfig.Validate(); err != nil {
+		return nil, err
+	}
+
+	if taskGroup == "" {
+		return &commandBlock{}, nil
+	}
+
+	tg := tc.taskConfig.Project.FindTaskGroup(taskGroup)
+	if tg == nil {
+		return nil, errors.Errorf("couldn't find task group '%s' in project '%s'", taskGroup, tc.taskConfig.Project.Identifier)
+	}
+	if tg.SetupGroup == nil {
+		return &commandBlock{}, nil
+	}
+
+	return &commandBlock{
+		block:       command.SetupGroupBlock,
+		commands:    tg.SetupGroup,
+		timeoutKind: setupGroupTimeout,
+		getTimeout:  tc.getSetupGroupTimeout(tg),
+		canFailTask: tg.SetupGroupCanFailTask,
+	}, nil
+}
+
+// getTeardownGroup returns the teardown group for a task group task.
+func (tc *taskContext) getTeardownGroup(taskGroup string) (*commandBlock, error) {
+	if err := tc.taskConfig.Validate(); err != nil {
+		return nil, err
+	}
+
+	if taskGroup == "" {
+		return &commandBlock{}, nil
+	}
+
+	tg := tc.taskConfig.Project.FindTaskGroup(taskGroup)
+	if tg == nil {
+		return nil, errors.Errorf("couldn't find task group '%s' in project '%s'", taskGroup, tc.taskConfig.Project.Identifier)
+	}
+	if tg.TeardownGroup == nil {
+		return &commandBlock{}, nil
+	}
+
+	return &commandBlock{
+		block:       command.TeardownGroupBlock,
+		commands:    tg.TeardownGroup,
+		timeoutKind: teardownGroupTimeout,
+		getTimeout:  tc.getTeardownGroupTimeout(tg),
+		canFailTask: false,
+	}, nil
+}
+
+// getTimeout returns a command block containing the timeout handler commands.
```

Comment:
Do we want to specify in the comment that this is for the task group callback timeout?